### ISSUE TITLE
Talawa-Admin Docker Support Issue

### ISSUE THREAD

**Issue by duplixx** [PalisadoesFoundation/talawa-admin#1075]
### Talawa-Admin Docker Support Issue:

**Title:**
Adding Docker Support to Talawa-Admin

**Description:**

**Problem:**
Talawa-Admin currently lacks Docker support, hindering its ease of deployment and scalability. Adding Docker support is essential to streamline the deployment process and make it more accessible to a broader audience.

**Solution:**
Implementing Docker support for Talawa-Admin would involve creating a Dockerfile and configuration to package the application into a Docker container. This containerization would simplify deployment, improve consistency across environments, and enhance scalability.

**Approach:**
To achieve Docker support for Talawa-Admin, we can follow these steps:
1. Create a Dockerfile specifying the necessary dependencies and configuration.
2. Configure the application to work seamlessly within a Docker container.
3. Document the steps for building and running Talawa-Admin using Docker in the project's README.

**Additional Context:**
Adding Docker support aligns with modern development practices, making it easier for developers to set up and run Talawa-Admin in various environments. Docker containers provide isolation, portability, and reproducibility, contributing to a more robust and user-friendly development and deployment experience.

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by duplixx** [comment]
Hey @palisadoes please assign me this issue.
Thanks !!

**Comment by noman2002** [comment]
@duplixx Do we even need docker on the admin? If yes, please explain why we need it.

**Comment by duplixx** [comment]
Hey @noman2002 , I strongly believe that integrating Docker into Talawa-Admin is essential. It effectively addresses significant challenges faced by developers and contributors.

- **Reduces Setup Issues for Contributors:**
  - Docker ensures a standardized environment, minimizing setup issues for contributors.
  - Facilitates a more consistent development environment for everyone involved.

- **Eliminates "It Works on My Machine" Problem:**
  - Docker eliminates the common challenge of code working differently on different machines.
  - Developers can confidently share and collaborate on the same codebase without encountering environment-related discrepancies.

- **Ensures Consistency Across Environments:**
  - Talawa-Admin benefits from consistent environments across development, testing, and production.
  - Docker ensures that the application behaves uniformly, mitigating issues related to environment variations.

- **Confident Collaboration:**
  - Developers can work on the same codebase without concerns about environment differences.
  - Docker fosters confident collaboration, enabling a smoother development process.

In summary, leveraging Docker in Talawa-Admin significantly improves collaboration, consistency, and reduces common challenges associated with environment setups.

**Comment by aashimawadhwa** [comment]
@kb-0311 @beingnoble03 @DMills27 would like to hear your thoughts on this.

**Comment by DMills27** [comment]
Docker is essentially being used as a stop gap, at the moment, until we can put in place a more sustainable means to achieve reproducibility. I propose have a single Docker image that generates the API, Admin and Talawa Mobile Frontend. My reasoning is that it would incur additional complexity to create separate Docker images for these and then connect them; it is also inefficient in terms of computational resources.

**Comment by palisadoes** [comment]
@DMills27 

1. You mean API and the web Admin/User portal correct?
2. How would we best approach this if we are using separate repos for the API and Admin?

**Comment by aarishshahmohsin** [comment]
I could make a dockerfile for Talawa-Admin. I there is a need, Please assign this issue to me.

**Comment by duplixx** [comment]
> @DMills27
> 
> 1. You mean API and the web Admin/User portal correct?
> 2. How would we best approach this if we are using separate repos for the API and Admin?

In my view we should follow this approach

### 1. Dockerfile for Each Component
   - Create a Dockerfile for API and Admin.
   - Specify base images with necessary dependencies.

### 2. Docker Compose for Multi-Container Setup
   - Use Docker Compose for a multi-container setup.
   - Define services, networks, and volumes in `docker-compose.yml`.

### 3. Service Orchestration
   - Define dependencies between services.
   - Use container names for inter-service communication.

**Comment by noman2002** [comment]
@DMills27 @palisadoes Is this issue required or should we close this for now ??

**Comment by EshaanAgg** [comment]
It would only make sense for this issue if we publish the `Talawa-API` as a docker image on DockerHub, and then use it to setup the admin via Compose. Otherwise there is no need for the same as of now IMO.

**Comment by noman2002** [comment]
Thanks @EshaanAgg . 
Closing this for now.

**Comment by EshaanAgg** [comment]
We are yet to make a docker image for Talawa APi. It would be apt to reopen the same after it.

**Comment by palisadoes** [comment]
> We are yet to make a docker image for Talawa APi. It would be apt to reopen the same after it. 

The API got its docker instance created about 2 weeks ago. So we should reopen this issue.

**Comment by Min2who** [comment]
can i get assigned?

**Comment by Min2who** [comment]
@EshaanAgg @noman2002 is the talawa image is published on dockerhub i cant see it?

**Comment by Min2who** [comment]
> The API got its docker instance created about 2 weeks ago. So we should reopen this issue.

@palisadoes @noman2002 i cant find the docker instance

**Comment by shankeleven** [comment]
@EshaanAgg  please assign this issue to me, would love to work

**Comment by Devesh326** [comment]
@palisadoes can i work on this issue!

**Comment by Devesh326** [comment]
Hey @aashimawadhwa and @palisadoes 

I'm currently working on dockerizing Talawa Admin. To set up the Docker Compose file, I need the Docker image for Talawa API, but I couldn't find it anywhere. 

Could you please guide me in the right direction or provide information about where I can obtain the Talawa API Docker image?

Thanks!

**Comment by palisadoes** [comment]
1. You'll need to assume that the API is up and running.
2. You will also need to update the setup script to accommodate the Docker support you'll be adding
    - https://github.com/PalisadoesFoundation/talawa-admin/issues/1233

**Comment by Devesh326** [comment]
Okay, thanks!

**Comment by Devesh326** [comment]
> 1. You'll need to assume that the API is up and running.
> 2. You will also need to update the setup script to accommodate the Docker support you'll be adding
>    
>    * [Create a Setup Script for the `.env` configuration file #1233](https://github.com/PalisadoesFoundation/talawa-admin/issues/1233)


Hi @palisadoes,

I have created a PR to add dockerization support for talawa admin.
Will update the setup script once that pr is merged to avoid conflicts.

**Comment by palisadoes** [comment]
Unassigning due to inactivity

**Comment by varshith257** [comment]
I would like to work on it

**Comment by varshith257** [comment]
Reviewing the previously attempted work

**Comment by varshith257** [comment]
@palisadoes Can I raise an issue for potential improvements to `Talawa-API`'s Docker support? It's mandatory for this issue and also i need to publish the `Talawa-Api` docker image on DockerHub.

**Comment by varshith257** [comment]
> It would only make sense for this issue if we publish the `Talawa-API` as a docker image on DockerHub, and then use it to setup the admin via Compose. Otherwise there is no need for the same as of now IMO.

This is what I intended to do

**Comment by varshith257** [comment]
@palisadoes Can i have a update on this?

**Comment by palisadoes** [comment]
> @palisadoes @Cioppolo14 @noman2002  Can I raise an issue for potential improvements to `Talawa-API`'s Docker support? It's mandatory for this issue and also i need to publish the `Talawa-Api` docker image on DockerHub.

We already have a docker API issue

**Comment by vasujain275** [comment]
@varshith257 

1. You only need to create a Dockerfile for Talawa-Admin in this issue.
2. The Dockerfile should use multi-stage Docker builds so that it can have a small image size and be more efficient.
3. Once the Dockerfile is merged, this issue will be closed, and we will need to raise a new issue to create a GitHub workflow to upload the Talawa Admin image to Docker Hub.
4. Regarding the Talawa API Docker Image to Docker Hub, it has been worked on completely as I stated [here](https://github.com/PalisadoesFoundation/talawa-api/issues/1751#issuecomment-1957364160).
5. If you want, I can handle the Talawa API Multi-Staged Build if it's too overwhelming right now. You can focus on the Talawa Admin Docker Image only.

**Comment by vasujain275** [comment]
Can I work on this?

**Comment by Cioppolo14** [comment]
Unassigning due to no activity or open PR.

**Comment by varshith257** [comment]
@Cioppolo14 Waiting to upload the docker-api issue to be merged and to be upload it to Docker-Hub as it is dependent on the Talawa-Api Docker image. As already same has been discussed in Talawa-Api for DockerHub Credentials

Can you please reassign me this issue?

**Comment by vasujain275** [comment]
> @Cioppolo14 Waiting to upload the docker-api issue to be merged and to be upload to Docker-Hub as this issue is dependent on the Talawa-Api Docker image. As already same has been discussed in Talawa-Api for DockerHub Credentials
> 
> Can you please reassign me this issue?
> 
> Almost done with this work but waiting to be the PR of talawa-api docker issue to be merged and same to be uploaded to Docker-Hub

How is the docker hub talawa api related to the docker file of talawa admin? Correct me if I am wrong but in the discussions we had, we do not intend to  upload the talawa admin docker image to the docker hub. I have started work on this too locally.

**Comment by varshith257** [comment]
> @palisadoes @Cioppolo14 @noman2002 Can I raise an issue for potential improvements to `Talawa-API`'s Docker support? It's mandatory for this issue and also i need to publish the `Talawa-Api` docker image on DockerHub.

@Cioppolo14 Same has been stated before and Sorry not to be commented about progress of work. Just now came here to update about the progress of work but unassigned

**Comment by varshith257** [comment]
> It would only make sense for this issue if we publish the `Talawa-API` as a docker image on DockerHub, and then use it to setup the admin via Compose. Otherwise there is no need for the same as of now IMO.

@vasujain275 Same as above. Then how could we configure API service for talawa-admin without image of talawa-api importing here?

**Comment by vasujain275** [comment]
> > It would only make sense for this issue if we publish the `Talawa-API` as a docker image on DockerHub, and then use it to setup the admin via Compose. Otherwise there is no need for the same as of now IMO.
> 
> @vasujain275 Same as above. Then how could we configure API service for talawa-admin without image of talawa-api importing here?

Why do we need talawa api image imported here, as this issue states we need a way for developers to locally run docker container of talawa admin for easy local devlopment. The API link will be provided by devloper setting this image locally only.

**Comment by varshith257** [comment]
How can talawa-admin run without talawa-api running? The intention of docker support to talawa-admin is to setup admin from docker-compose and doesn't to rely on the talwa-api to be started locally it doesn't make any sense of it.

**Comment by vasujain275** [comment]
@varshith257 As peter suggested we don't want to much integration of talawa api and talawa admin right now as they are very dynamic in nature and we came to conclusion to make them seprate docker instances. In future when talawa applications are less dynamic and stable we can move to a closer intergration. So, we decided on an 2 independent docker image for talawa admin one for multi-stage, light weight and one image for dev environment with code attached as volumes. Please refer to above disscussion in the pr for more contex.

**Comment by varshith257** [comment]
@palisadoes Can you provide your suggestions here what exactly need to be added now just with docker support of talawa-admin and assume locally talawa-api is up.

If it is the case, done with the work and ready to raise a PR

OR

The same  intended by @DMills27 and @EshaanAgg in above comments to support talawa admin- setup  without again relying to up talwa-api manually

**Comment by vasujain275** [comment]
@varshith257 You don't need to worry about your work being a waste. I was just talking about the approach we prefer right now for talawa admin docker image according to previous conversations. If you are ready with the pr there is no sense for me to work the same thing all over again

**Comment by varshith257** [comment]
> @palisadoes Can you please provide your suggestions here what exactly need to be added now just with docker support of talawa-admin and assume locally talawa-api is up.
> 
> If it is the case, done with the work and ready to raise a PR
> 
> OR
> 
> The same  intended by @DMills27 and @EshaanAgg in above comments to support talawa admin- setup  without again relying to up talwa-api manually

@palisadoes @Cioppolo14

**Comment by varshith257** [comment]
> Please work with @noman2002 on this

Okay! @Cioppolo14 Can I get reassigned this issue?

**Comment by vasujain275** [comment]
@varshith257 If you don't mind can I work on this? I have a good approach in mind, did some research about it too

**Comment by varshith257** [comment]
> @varshith257 If you don't mind can I work on this? I have a good approach in mind, did some research about it too

But I have already done with work when I got assigned this issue and worked it for a week. In thought API docker issue is dependent on this as intended by Eshaan I have waited the API docker image to be uploaded in Docker Hub.

**Comment by Cioppolo14** [comment]
This is issue has a complicated assignment history.  In the future, if you feel you are waiting for answers, please keep the issue active and/or track down answers in Slack so work can progress. I will not change the assignee right now.

**Comment by vasujain275** [comment]
Here is the approach I am taking, commenting for refrence in future - 
1. I will create two docker files - 
     - Dockerfile.dev for local development which also have code attached as volume for hot refresh feature.
     - Dockerfile.prod for production which uses multi stage docker files and serve pre build js files using final stage node:alpine image.
2. Once these two are implemented I will docker compose for the respective dockerfiles.
3. Then I will add npm script to `package.json` for `npm run dx-dev` an `npm run dx-prod` for easily startiing the containers.
4. Finally, I will modify the Installation docs for the same.

**Comment by vasujain275** [comment]
I have raised the PR for the Docker Support in Talawa Admin, here is the Summary of the changes - 

**1. Introduction of Docker Support:**
- This PR adds Docker support to Talawa Admin.
- This enhancement aims to facilitate a smoother developer experience, especially for individuals contributing to Talawa API and Talawa App.

**2. Consideration of Docker Methods:**
- Initially, the idea was to implement two Docker methods - dev and production.
- However, upon scrutinizing the codebase, it was discovered that we are utilizing `Create React App` and react scripts for the build process.
- The current build process is evidently not optimized for either production-ready minified binaries or Docker hot reloading.

**3. Proposal for Build Process Improvement:**
- To address this limitation, I think we should migrate to a more suitable build process like Vite.
- This migration is necessary to enable efficient building of both minified production-ready build files and Docker hot reloading.

**4. Planned Actions:**
- After this PR is merged, I will raise an issue to migrate from React scripts to the Vite build process, we can have further discussion on build process there.
- This action will lay the groundwork for future improvements in better and improved Docker support once the Build Process is sorted.

**5. Current Status:**
- Currently, separate dev and prod Docker methods have not been implemented due to the inadequacy of the existing build process.
- The current build process cannot produce minified production-ready build files, and a dev hot reloading Docker setup would be impractical due to the same issue.


@kb-0311 @noman2002 @palisadoes I highly recommand to migrate to VIte Build process, it will make our build process so much efficient, faster, simplified. I would appreciate your insights on this.

**Comment by vasujain275** [comment]
@Cioppolo14 Any update on pr review?

**Comment by Cioppolo14** [comment]
@vasujain275 Sorry for the delay, I am back on it. I will follow up soon. Feel free to find me on slack too.

**Comment by Cioppolo14** [comment]
I am working on the review.  Starting with API and working my way through.

**Comment by VanshikaSabharwal** [comment]
@palisadoes  I understand that the primary goal of the PR is to ensure the correct startup of Talawa-Admin within a Docker container, which is a valuable addition to the workflow.

**Comment by VanshikaSabharwal** [comment]
@palisadoes  @DMills27  @noman2002  @varshith257 I'm open to collaborating with you and other contributors to implement Docker support for Talawa-Admin. We can discuss specific strategies and potential challenges.

Please let me know if you have any questions or require further assistance.

**Comment by VanshikaSabharwal** [comment]
Is this Dockerfile GTG or should i add something more. 
![image](https://github.com/user-attachments/assets/89258d15-7140-42fd-9c11-b953ee97e8e9)

**Comment by varshith257** [comment]
@VanshikaSabharwal You can refer the previous attempt made which is close to what we need

https://github.com/PalisadoesFoundation/talawa-admin/pull/1726

**Comment by VanshikaSabharwal** [comment]
> @VanshikaSabharwal You can refer the previous attempt made which is close to what we need
> 
> #1726

Okay i get it

**Comment by VanshikaSabharwal** [comment]
I was facing this issue while running docker file locally for talawa-admin. 
![image](https://github.com/user-attachments/assets/b67f0a51-2071-4593-9c75-251dadbfe8e1)
. So i had to  modify this file vite.config.ts file. 
Before.
![image](https://github.com/user-attachments/assets/6da9eed0-30bf-4209-9daf-5062c23cd2a9)

After
![image](https://github.com/user-attachments/assets/e80e7d3d-32c5-4650-accc-a398e47539a8)
. After making above changes i am now able to run Docker file for talawa-admin. 
![image](https://github.com/user-attachments/assets/8714f449-6a4e-4d7a-b0b1-9d0f644551b0)
![image](https://github.com/user-attachments/assets/b0ce84de-a0a0-4f23-85ac-e9a66744f1f6)

SO should i raise PR now??

**Comment by palisadoes** [comment]
If the problem is fixed, please open the PR

**Comment by palisadoes** [comment]
During the week of November 11, 2024 we will start a code freeze on the `develop` branches in Talawa, Talawa Admin and Talawa-API. 

We have completed a project to convert the Talawa-API backend to use PostgreSQL. Work will then begin with us merging code in the `develop` branches to a new `develop-postrgres` branch in each repository.

Planning activities for this will be managed in our #talawa-projects slack channel. A GitHub project will be created to track specially labeled issues. We completed a similar exercise last year using a similar methodology. 

Starting November 12, California time no new PRs will be accepted against the `develop` branch. They must be applied to the `develop-postrgres` branch. 

There are some GSoC project features that will need to be merged into develop. These will be the only exceptions.

This activity and the post GSoC 2024 start date was announced in our #general Slack channel last month as a pinned post.
